src/main/java/com/puppycrawl/tools/checkstyle/site/ParentModuleMacro.java

```diff
@@ -0,0 +1,165 @@
+     * @return the parent module name as a string.
+     * @throws MacroExecutionException if the parent module cannot be found.
+     */
+    private static String getParentModule(Class<?> moduleClass, String moduleName)
+                throws MacroExecutionException {
+        String parentModuleName = "";
+        Class<?> parentClass = moduleClass.getSuperclass();
+        while (parentClass != null && parentModuleName.isEmpty()) {
+            if (parentClass == AbstractCheck.class
+                    || parentClass == TreeWalkerFilter.class) {
+                parentModuleName = "TreeWalker";
+            }
+            else if (parentClass == AbstractFileSetCheck.class
+                    || parentClass == Filter.class) {
+                parentModuleName = "Checker";
+            }
+            else {
+                parentClass = parentClass.getSuperclass();
+            }
+        }
+
+        if (parentModuleName.isEmpty()) {
+            final String message = String.format(Locale.ROOT,
+                    "Failed to find parent module for %s", moduleName);
+            throw new MacroExecutionException(message);
+        }
+
+        return parentModuleName;
+    }
+
+    /**
+     * Returns relative link to the parent module for the given module class.
+     *
+     * @param parentModule parent module name.
+     * @param moduleName the module name we are looking for the parent of.
+     * @return relative link to the parent module.
+     * @throws MacroExecutionException if link to the parent module cannot be constructed.
+     */
+    private static String getLinkToParentModule(String parentModule, String moduleName)
+            throws MacroExecutionException {
+        final Path templatePath = getTemplatePath(moduleName);
+        final Path templatePathParent = templatePath.getParent();
+        if (templatePathParent == null) {
+            throw new MacroExecutionException("Failed to get parent path for " + templatePath);
+        }
+        return templatePathParent
+                .relativize(Paths.get("src", "xdocs", "config.xml"))
+                .toString()
+                .replace(".xml", ".html")
+                + "#" + parentModule;
+    }
+
+    /**
+     * Returns path to the template for the given module name or throws an exception if the
+     * template cannot be found.
+     *
+     * @param moduleName the module whose template we are looking for.
+     * @return path to the template.
+     * @throws MacroExecutionException if the template cannot be found.
+     */
+    private static Path getTemplatePath(String moduleName) throws MacroExecutionException {
+        try {
+            final String fileNamePattern = ".*[\\\\/]"
+                    + moduleName.toLowerCase(Locale.ROOT) + "\\..*";
+            return SiteUtil.getXdocsTemplatesFilePaths()
+                .stream()
+                .filter(path -> path.toString().matches(fileNamePattern))
```

Comment:
I couldn't import `import org.checkerframework.checker.regex.RegexUtil;` so I'm suppressing the checker violation
https://github.com/checkstyle/checkstyle/actions/runs/5752608316/job/15593968255?pr=13435#step:6:286

```
   <checkerFrameworkError unstable="false">
    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/site/ParentModuleMacro.java</fileName>
    <specifier>argument</specifier>
    <message>incompatible argument for parameter regex of matches.</message>
    <lineContent>.filter(path -&gt; path.toString().matches(fileNamePattern))</lineContent>
    <details>
      found   : String
      required: @Regex String
    </details>
  </checkerFrameworkError>
```

https://checkerframework.org/releases/1.9.7/tutorial/webpages/user-input-cmd.html